### PR TITLE
Code Coverage Feature

### PR DESCRIPTION
**Pull request by steuhs**
calculate, summarize and report code coverage through covbot and testgrid

This PR will be closed:
Re-organized the PR into three commits here #9313
Design doc in this PR #9302

### CODE REVIEW

**Comment by cjwagner** [comment]
I'll do another review of this tomorrow. It'd be great to get some other eyes on this though.
/cc @fejta @stevekuznetsov @amwat

**Comment by stevekuznetsov** [comment]
Do we have a design doc to come along with +4k LOC change? :)

File: coverage/artifacts/artifacts.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+	"path"
+)
+
+const (
+	CovProfileCompletionMarker = "profile-completed"
+	JunitXmlForTestgrid        = "junit_bazel.xml"
+	LineCovFileName            = "line-cov.html"
+)
+
+type Intf interface {
```
[the remark this answers is not in the document]
Comment (steuhs) [reply coverage/artifacts/artifacts.go]: It's the interface shared by Artifacts, LocalArtifacts & GcsArtifacts

[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply coverage/artifacts/artifacts.go]: `Intf` is not a useful name unfortunately -- name the interface to describe what it function it has, such that any `struct` implementing the interface also has. See names like `io.Reader`

File: coverage/calc/calc.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	"k8s.io/test-infra/coverage/artifacts"
+	"log"
+	"os"
+)
+
+// CovList read profiling information from reader and constructs CoverageList.
+// If called in presubmit, it also creates a filtered version of profile,
+// that only includes files in corresponding github commit,
+// less those files that are excluded from coverage calculation
+func CovList(f *artifacts.ProfileReader, keyProfileFile *os.File,
```
[the remark this answers is not in the document]
Comment (steuhs) [reply coverage/calc/calc.go]: it actually needs to be a WriteCloser

File: coverage/calc/base.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+func (c *Coverage) Percentage() string {
+	ratio, err := c.Ratio()
+	if err == nil {
+		return str.PercentStr(ratio)
+	}
+
+	return "N/A"
+}
+
+// PercentageForTestgrid returns the percentage of statements covered
```
[the remark this answers is not in the document]
Comment (steuhs) [reply coverage/calc/base.go]: Percentage() is the string used for internal logging and display on covbot, which may not have the same format used by testgrid. Even though currently both keep 1 decimal place.

File: coverage/artifacts/artifacts.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+func New(directory string, profileName string, keyProfileName string,
+	covStdoutName string) *Artifacts {
+	return &Artifacts{
+		directory,
+		profileName,
+		keyProfileName,
+		covStdoutName}
+}
+
+func (arts *Artifacts) SetDirectory(dir string) {
```
[the remark this answers is not in the document]
Comment (steuhs) [reply coverage/artifacts/artifacts.go]: so that we control how the field can be set/get

File: coverage/artifacts/local.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	return arts.profileName
+}
+
+// KeyProfileCreator creates a key profile file that will be used to hold a
+// filtered version of coverage profile that only stores the entries that
+// will be displayed by line coverage tool
+func (arts *LocalArtifacts) KeyProfileCreator() *os.File {
+	keyProfilePath := arts.KeyProfilePath()
+	keyProfileFile, err := os.Create(keyProfilePath)
+	log.Printf("os.Create(keyProfilePath)=%s", keyProfilePath)
```
[the remark this answers is not in the document]
Comment (steuhs) [reply coverage/artifacts/local.go]: I think it helps to print the log here

File: coverage/calc/base.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+func (blk *codeBlock) addToFileCov(coverage *Coverage) {
+
+	coverage.nAllStmts += blk.numStatements
+	if blk.coverageCount > 0 {
+		coverage.nCoveredStmts += blk.numStatements
+	}
+}
+
+// add blk Coverage to file group Coverage; return true if the row is concerned
+func updateConcernedFiles(concernedFiles *map[string]bool, filePath string,
```
[the remark this answers is not in the document]
Comment (steuhs) [reply coverage/calc/base.go]: do you mean to use 'map' instead?

File: coverage/artifacts/artifacts.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (steuhs) [reply coverage/artifacts/artifacts.go]: Done.

File: coverage/artifacts/artifacts.go [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+type Artifacts struct {
+	directory      string
+	profileName    string
+	keyProfileName string
+	covStdoutName  string
+}
+
+func New(directory string, profileName string, keyProfileName string,
```
[the remark this answers is not in the document]
Comment (steuhs) [reply coverage/artifacts/artifacts.go]: Done.

**Comment by BenTheElder** [comment]
/uncc
Let me know when this is ready for futher review.

File: coverage/artifacts/profiling.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+	cmdArgs = append(cmdArgs, []string{"-covermode=count",
+		"-coverprofile", localArts.ProfilePath()}...)
+
+	log.Printf("go cmdArgs=%v\n", cmdArgs)
+	cmd := exec.Command("go", cmdArgs...)
+
+	goTestCoverStdout, errCmdOutput := cmd.Output()
+
+	if errCmdOutput != nil {
+		logUtil.LogFatalf("Error running 'go test -coverprofile ': error='%v'; stdout='%s'; stderr='%v'",
```
[the remark this answers is not in the document]
Comment (steuhs) [reply coverage/artifacts/profiling.go]: What would be a better alternative? It's hard to avoid multi-line output here

File: coverage/calc/base.go [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+func (c *Coverage) PercentageForTestgrid() string {
+	ratio, err := c.Ratio()
+	if err == nil {
+		return str.PercentageForTestgrid(ratio)
+	}
+
+	return ""
+}
+
+func (c *Coverage) Ratio() (ratio float32, err error) {
```
[the remark this answers is not in the document]
Comment (steuhs) [reply coverage/calc/base.go]: Done.

File: coverage/calc/base.go [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+
+func (c *Coverage) SetLineCovLink(link string) {
+	c.lineCovLink = link
+}
+
+// IsCoverageLow checks if the coverage is less than the threshold.
+func (c *Coverage) IsCoverageLow(covThresholdInt int) bool {
+	covThreshold := float32(covThresholdInt) / 100
+	ratio, err := c.Ratio()
+	if err == nil {
```
[the remark this answers is not in the document]
Comment (steuhs) [reply coverage/calc/base.go]: Done.

File: coverage/calc/calc.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (steuhs) [reply coverage/calc/calc.go]: Done.

File: coverage/artifacts/artsTest/artsTest.go [hunk 11]
```diff
@@ -0,0 +1,39 @@
+// package artsTest stores artifacts functions for tests,
+// used by other packages
+package artsTest
+
+import (
+	"k8s.io/test-infra/coverage/artifacts"
+	"k8s.io/test-infra/coverage/test"
+)
+
+type LocalArtifacts = artifacts.LocalArtifacts
```
[the remark this answers is not in the document]
Comment (steuhs) [reply coverage/artifacts/artsTest/artsTest.go]: Done.

File: coverage/artifacts/local_test.go [hunk 12]
```diff
@@ -0,0 +1,47 @@
+package artifacts
+
+import (
+	"bufio"
+	"fmt"
+	"k8s.io/test-infra/coverage/logUtil"
```
[the remark this answers is not in the document]
Comment (steuhs) [reply coverage/artifacts/local_test.go]: Done.

File: coverage/artifacts/profiling.go [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+func NewProfileReader(reader io.ReadCloser) *ProfileReader {
+	return &ProfileReader{reader}
+}
+
+// runProfiling writes coverage profile (&its stdout) by running go test on
+// target package
+func runProfiling(covTargets []string, localArts *LocalArtifacts) {
+	log.Println("\nStarts calc.runProfiling(...)")
```
[the remark this answers is not in the document]
Comment (steuhs) [reply coverage/artifacts/profiling.go]: Done.

File: coverage/artifacts/local.go [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+
+	return keyProfileFile
+}
+
+// ProduceProfileFile produce coverage profile (&its stdout) by running go test on target package
+// for periodic job, produce junit xml for testgrid in addition
+func (arts *LocalArtifacts) ProduceProfileFile(covTargetsStr string) {
+	// creates artifacts directory
+	log.Printf("mkdir -p %s\n", arts.directory)
+	cmd := exec.Command("mkdir", "-p", arts.directory)
```
[the remark this answers is not in the document]
Comment (steuhs) [reply coverage/artifacts/local.go]: Done.

File: coverage/artifacts/artifacts.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (cjwagner) [reply coverage/artifacts/artifacts.go]: You aren't using the getters and setters to control access or modify behavior though. `arts.directory` is just assigned in the setter and read in the getter so we could just export `arts.Directory` instead.

File: coverage/calc/base.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (cjwagner) [reply coverage/calc/base.go]: Yes, `map`s are already reference types.

File: coverage/calc/base.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply coverage/calc/base.go]: Let's make the distinction once we need to, then, instead of optimizing for the future that may never happen :)

**Comment by stevekuznetsov** [comment]
@steuhs would still like to see a design document with high-level vision, what we are achieving, acceptance criteria, who are the stakeholders, how can we measure success...

**Comment by steuhs** [comment]
@stevekuznetsov @cjwagner I uploaded the design doc in docs/design. Please take a look. I am working on adding the acceptance criteria. Otherwise it should have all the needed parts

File: coverage/docs/design.md [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+Prow has handlers for different github events. We add a pre-submit prow job that is triggered by any new commit to a PR to run test coverage on the new build to compare it with the master branch and previous commit for pre-submit coverage. We add a post-submit prow job that is triggered by merge events, to run test coverage on the nodes of master branch. Test coverage data on the master branch is supplied to TestGrid for displaying the coverage change over time, as well as serve as the basis of comparison for pre-submit coverage mentioned in the pre-submit senario.
+
+Here is the step-by-step description of the pre-submit and post-submit workflows
+
+##Pre-submit workflow
+Runs code coverage tool to report coverage change in a new commit or new PR
+1. Developer submit new commit to an open PR on github
+2. Matching pre-submit prow job is started 
+3. Test coverage profile generated
+4. Calculate coverage change against master branch. Compare the coverage file generated in this cycle against the most recent successful post-submit build. Coverage file for post-submit commits were generated in post-submit workflow
```
Comment (stevekuznetsov) [inline coverage/docs/design.md]: What if the pull request is not targeting the `master` branch?

Comment (stevekuznetsov) [reply coverage/docs/design.md]: Where are data about historical results for other commit stored?

File: coverage/docs/design.md [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+5. Use PR data from github, git-attributes, as well as coverage change data calculated above, to produce a list of files that we care about in the line-by-line coverage report. produce line by line coverage html and add link to covbot report.
```
Comment (stevekuznetsov) [inline coverage/docs/design.md]: What is `covbot`?

File: coverage/docs/design.md [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+6. Let covbot post presubmit coverage on github, under that conversation of the PR.
```
Comment (stevekuznetsov) [inline coverage/docs/design.md]: Does this use the Prow `presubmit` flow? Why do we not report using `plank`?

File: coverage/docs/design.md [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+3. Test coverage profile generated
+4. Calculate coverage change against master branch. Compare the coverage file generated in this cycle against the most recent successful post-submit build. Coverage file for post-submit commits were generated in post-submit workflow
+5. Use PR data from github, git-attributes, as well as coverage change data calculated above, to produce a list of files that we care about in the line-by-line coverage report. produce line by line coverage html and add link to covbot report.
+6. Let covbot post presubmit coverage on github, under that conversation of the PR. 
+
+##Post-submit workflow
+Produces & stores coverage profile for later presubmit jobs to compare against
+1. A PR is merged
+2. Post-submit prow job started
+3. Test coverage profile generated. Completion marker generated upon successful run.
```
Comment (stevekuznetsov) [inline coverage/docs/design.md]: Where is the profile stored? What is the marker used for?

File: coverage/docs/design.md [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+
+##Periodical workflow
```
Comment (stevekuznetsov) [inline coverage/docs/design.md]: Why is this required if postsubmits are running?

File: coverage/docs/design.md [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+Produces periodical coverage result as input for TestGrid
+1. Periodical prow job starts periodically based on the specification in prow job config
+2. Test coverage profile & metadata generated 
+3. Generate / store per-file coverage data
+  - Stores in the XML format that is used by TestGrid
+  
+##Prow Configuration File
+As mentioned earlier, we use configuration file to store repository specific information. Below is an example that contains the args that will be supplied to the coverage container
+```
+    spec:
```
Comment (stevekuznetsov) [inline coverage/docs/design.md]: You might want the rest of the Prow Job here

File: coverage/docs/design.md [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+2. It receives github related variables, such as pull request number & commit number, from Prow. Those variables are used as meta-data for the profiles. Metadata allows us to do presumbit coverage comparisons.
+3. It allows user to supply variables in prowjob configs. Those variables include directory to run test coverage, file filters and threshold for desired coverage level.  
+
+Prow has handlers for different github events. We add a pre-submit prow job that is triggered by any new commit to a PR to run test coverage on the new build to compare it with the master branch and previous commit for pre-submit coverage. We add a post-submit prow job that is triggered by merge events, to run test coverage on the nodes of master branch. Test coverage data on the master branch is supplied to TestGrid for displaying the coverage change over time, as well as serve as the basis of comparison for pre-submit coverage mentioned in the pre-submit senario.
+
+Here is the step-by-step description of the pre-submit and post-submit workflows
+
+##Pre-submit workflow
+Runs code coverage tool to report coverage change in a new commit or new PR
```
[the remark this answers is not in the document]
Comment (steuhs) [reply coverage/docs/design.md]: explained at the top of the doc

File: coverage/docs/design.md (hunk 15 above)
Comment (steuhs) [reply coverage/docs/design.md]: It doesn't have to be master branch - will change the wording here.  Data will be stored in artifacts, similar to that of other prow jobs

File: coverage/docs/design.md [hunk 22]
```diff
@@ -0,0 +1,64 @@
+#Overview
+Code coverage tool has two major features.
+1. As a pre-submit tool, it runs code coverage on every single commit to Github and reports coverage change back to the PR as a comment by a robot account. It also has the ability to block a PR from merging if coverage falls below threshold
```
[the remark this answers is not in the document]
Comment (steuhs) [reply coverage/docs/design.md]: > How are files chosen for the coverage denominator?
there is no cross file summary for pre-submit. Coverage reported is per file only
 
The rest of the questions are answered in updated doc

File: coverage/docs/design.md [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+##Background - Prow
+Prow is a system that handles github events and commands and allow you to perform actions. It was originally built for kubernetes, but now is extended to other teams wanting to use it as well. cmd/hook is the main entry point for Prow and listens to the github events. Prow provides two ways to handle events:
+1. Prow Jobs: Jobs that perform simple actions when certain events occur. Can only report the result as PASS or FAIL. For eg. running tests
+2. Plugins: Some logic that perform more complicated actions like talking to external service. Can report any kind of status. For eg. Golint plugin that checks out code from github and performs linting. Plugins can be:
+  - Internal: Live within the cmd/hook binary.
+  - External: Live as a separate binary. Events are forwarded to these by cmd/hook.
+
+#Design of Test Coverage Tool
+We pack the test coverage feature in a container, that is triggered by prow. The feature takes input from three sources
+1. It runs test coverage profiling on target repository. Prow clones the target repository as the current working directory for the container.
```
[the remark this answers is not in the document]
Comment (steuhs) [reply coverage/docs/design.md]: It is running as a Prow Job. I am not sure what do you mean by "using the pod utilities"

File: coverage/docs/design.md (hunk 15 above)
Comment (steuhs) [reply coverage/docs/design.md]: Done.

File: coverage/docs/design.md (hunk 16 above)
Comment (steuhs) [reply coverage/docs/design.md]: Done.

File: coverage/docs/design.md (hunk 20 above)
Comment (steuhs) [reply coverage/docs/design.md]: Done.

**Comment by steuhs** [comment]
@stevekuznetsov @cjwagner updated doc and addressed all comments on doc

File: coverage/docs/design.md [hunk 24]
```diff
@@ -0,0 +1,113 @@
+#Overview
+This code coverage tool has two major features.
+1. As a pre-submit tool, it runs code coverage on every single commit to Github and reports coverage change back to the PR as a comment by a robot account. It also has the ability to block a PR from merging if coverage falls below threshold
```
Comment (stevekuznetsov) [inline coverage/docs/design.md]: I still think this needs development. What is being built? Here are my questions:

 - what is a `"pre-submit tool"`? Are you building something that specifically runs pre-submit? Are you just building a code coverage measurement tool? Split the runtime configuration out from _what_ you are building
 - the comments on running for every single commit and reporting and blocking merges -- are you building that? Are you using Prow's built-in job triggering and reporting? Write about what _your_ work will be doing

File: coverage/docs/design.md [hunk 25]
```diff
@@ -0,0 +1,113 @@
...
+2. As a post-submit / periodical running job, it reports on TestGrid to show users how coverage changes over time.
+
+##Users
+The presubmit tool is intended for a developer to see the impact on code coverage of his/her commit. It can also be used by repo managers to block any PR from merging if the coverage falls under customized threshold.
+
+The periodical testgrid report is for repo managers and/or test infra team to monitor code coverage stats over time.
+
+##Limitation
+As of now, the code coverage tool only collect code coverage for Go files. The support for more programming languages may be added later
```
Comment (stevekuznetsov) [inline coverage/docs/design.md]: This is pretty huge -- are you designing explicitly with the hope to add more languages later? Is that necessary at all? Are you making assumptions about Go in your implementation that would make support for other languages hard? Take a strong stance on this.

File: coverage/docs/design.md [hunk 26]
```diff
@@ -0,0 +1,113 @@
...
+
+##Background - Prow
```
Comment (stevekuznetsov) [inline coverage/docs/design.md]: Not sure this is necessary. You look to be developing a standalone tool that may or may not be run inside of a Prow job.

File: coverage/docs/design.md [hunk 27]
```diff
@@ -0,0 +1,113 @@
...
+Prow is a system that handles github events and commands and allow you to perform actions. It was originally built for kubernetes, but now is extended to other teams wanting to use it as well. cmd/hook is the main entry point for Prow and listens to the github events. Prow provides two ways to handle events:
+1. Prow Jobs: Jobs that perform simple actions when certain events occur. Can only report the result as PASS or FAIL. For eg. running tests
+2. Plugins: Some logic that perform more complicated actions like talking to external service. Can report any kind of status. For eg. Golint plugin that checks out code from github and performs linting. Plugins can be:
+  - Internal: Live within the cmd/hook binary.
+  - External: Live as a separate binary. Events are forwarded to these by cmd/hook.
+
+#Design of Test Coverage Tool
```
Comment (stevekuznetsov) [inline coverage/docs/design.md]: There's a lot here about specifics for Prow -- why is any of that necessary? If I run this tool myself in my local directory, would it work? The way that Prow runs these jobs has been specifically designed so that the jobs themselves (and thereby your tool) should not need to know any Prow specifics to run, and should not really even care whether it is Prow that triggered it.

File: coverage/docs/design.md [hunk 28]
```diff
@@ -0,0 +1,113 @@
...
+##Periodical workflow
+Produces periodical coverage result as input for TestGrid
+1. Periodical prow job starts periodically based on the specification in prow job config
+2. Generate coverage profile
+3. Generate / store per-file coverage data
+  - Stores in the XML format, that is used by TestGrid, and dump it in prow artifacts directory 
+  
+##Prow Configuration File
+As mentioned earlier, we use configuration file to store repository specific information. Below is an example that contains the args that will be supplied to the coverage container
+```$xslt
```
Comment (stevekuznetsov) [inline coverage/docs/design.md]: This  is YAML, not XML

File: coverage/docs/design.md [hunk 29]
```diff
@@ -0,0 +1,113 @@
+#Overview
+This code coverage tool has two major features.
+1. As a pre-submit tool, it runs code coverage on every single commit to Github and reports coverage change back to the PR as a comment by a robot account. It also has the ability to block a PR from merging if coverage falls below threshold
+2. As a post-submit / periodical running job, it reports on TestGrid to show users how coverage changes over time.
+
+##Users
+The presubmit tool is intended for a developer to see the impact on code coverage of his/her commit. It can also be used by repo managers to block any PR from merging if the coverage falls under customized threshold.
```
Comment (cjwagner) [inline coverage/docs/design.md]: This wording is confusing, it makes it sounds like you have implemented multiple different tools when in reality you have a single job that is running in multiple places.

Also like Steve noted, the job doesn't do anything to block a PR from merging, that functionality is provided by Tide. I think what you really mean is that job will fail if coverage falls below the configured threshold.

File: coverage/docs/design.md [hunk 30]
```diff
@@ -0,0 +1,113 @@
...
+2. Plugins: Some logic that perform more complicated actions like talking to external service. Can report any kind of status. For eg. Golint plugin that checks out code from github and performs linting. Plugins can be:
+  - Internal: Live within the cmd/hook binary.
+  - External: Live as a separate binary. Events are forwarded to these by cmd/hook.
+
+#Design of Test Coverage Tool
+We pack the test coverage feature in a container, that is triggered by prow as a prow job. There is a separate prow job configured for each of the following workflows (which is discussed in later sections): pre-submit, post-submit and periodic. 
+
+The feature takes input from three sources
+1. It runs test coverage profiling on target repository. Prow clones the target repository as the current working directory for the container.
+2. It receives github related variables, such as pull request number & commit number, from Prow. Those variables are used as meta-data for the profiles. Metadata allows us to do presumbit coverage comparisons.
```
Comment (cjwagner) [inline coverage/docs/design.md]: Do you really need the PR number and commit SHA to do coverage comparisons? I don't see why that would be needed given that the appropriate PR code is already checked out when the job runs.

File: coverage/docs/design.md [hunk 31]
```diff
@@ -0,0 +1,113 @@
...
+3. It allows user to supply variables in prowjob configs. Those variables include directory to run test coverage, file filters and threshold for desired coverage level.  
+
+Prow has handlers for different github events. We add a pre-submit prow job that is triggered by any new commit to a PR to run test coverage on the new build to compare it with the master branch and previous commit for pre-submit coverage. We add a post-submit prow job that is triggered by merge events, to run test coverage on the nodes of master branch. Test coverage data on the master branch is supplied to TestGrid for displaying the coverage change over time, as well as serve as the basis of comparison for pre-submit coverage mentioned in the pre-submit senario.
+
+Here is the step-by-step description of the pre-submit and post-submit workflows
+
+##Pre-submit workflow
+Runs code coverage tool to report coverage change in a new commit or new PR
```
Comment (cjwagner) [inline coverage/docs/design.md]: I think you just mean new or updated PR. You aren't calculating coverage results for each individual commit in the PR right?

File: coverage/docs/design.md [hunk 32]
```diff
@@ -0,0 +1,113 @@
...
+1. Developer submit new commit to an open PR on github
+2. Matching pre-submit prow job is started 
+3. Test coverage profile generated
+4. Calculate coverage changes. Compare the coverage file generated in this cycle against the most recent successful post-submit build. Coverage file for post-submit commits were generated in post-submit workflow
+5. Use PR data from github, git-attributes, as well as coverage change data calculated above, to produce a list of files that we care about in the line-by-line coverage report. produce line by line coverage html and add link to covbot report. Note that covbot is the robot github account used to report code coverage change results.
```
Comment (cjwagner) [inline coverage/docs/design.md]: This makes it sounds like @covbot specifically must be used. You should make it clear that this is an arbitrary bot account and explain what permissions are needed and how the bot account is used. 
It is probably worth explaining the bot token requirement in its own section.

File: coverage/docs/design.md [hunk 33]
```diff
@@ -0,0 +1,113 @@
+#Overview
```
Comment (cjwagner) [inline coverage/docs/design.md]: This doesn't mention the use of GCS for persistent storage. I think that is important to mention since the job fetches results from previous runs from GCS.

File: coverage/docs/design.md [hunk 34]
```diff
@@ -0,0 +1,113 @@
...
+As mentioned earlier, we use configuration file to store repository specific information. Below is an example that contains the args that will be supplied to the coverage container
+```$xslt
+  - name: pull-knative-serving-go-coverage
+    labels:
+      preset-service-account: "true"
+    agent: kubernetes
+    context: pull-knative-serving-go-coverage
+    always_run: true
+    rerun_command: "/test pull-knative-serving-go-coverage"
+    trigger: "(?m)^/test (all|pull-knative-serving-go-coverage),?(\\s+|$)"
```
Comment (cjwagner) [inline coverage/docs/design.md]: You can delete the agent, context, rerun_command, and trigger fields now.

File: coverage/docs/design.md [hunk 35]
```diff
@@ -0,0 +1,113 @@
...
+    ssh_key_secrets:
+    - ssh-knative
+    spec:
+      containers:
+      - image: gcr.io/knative-tests/test-infra/coverage:latest
+        imagePullPolicy: Always
+        command:
+        - "/coverage"
+        args:
+        - "--postsubmit-gcs-bucket=knative-prow"
```
Comment (stevekuznetsov) [inline coverage/docs/design.md]: Doesn't `plank` know this?

File: coverage/docs/design.md [hunk 36]
```diff
@@ -0,0 +1,113 @@
...
+        - "--postsubmit-job-name=post-knative-serving-go-coverage"
+        - "--artifacts=$(ARTIFACTS)" # folder to store artifacts, such as the coverage profile
+        - "--profile-name=coverage_profile.txt"
+        - "--cov-target=./pkg/" # target directory of test coverage
```
Comment (stevekuznetsov) [inline coverage/docs/design.md]: Can we use `go list` here? Do we expect this tool to get everything or just specific directories?

File: coverage/docs/design.md (hunk 35 above)
Comment (steuhs) [reply coverage/docs/design.md]: @cjwagner

File: coverage/docs/design.md (hunk 36 above)
Comment (steuhs) [reply coverage/docs/design.md]: just specific directory. We should allow each repo to specify the directory they care about for code coverage

File: coverage/docs/design.md (hunk 34 above)
Comment (steuhs) [reply coverage/docs/design.md]: Done.

File: coverage/docs/design.md (hunk 18 above)
Comment (steuhs) [reply coverage/docs/design.md]: Done.

File: coverage/docs/design.md (hunk 19 above)
Comment (steuhs) [reply coverage/docs/design.md]: Done.

File: coverage/docs/design.md (hunk 36 above)
Comment (steuhs) [reply coverage/docs/design.md]: just the directories supplied here

File: coverage/docs/design.md (hunk 33 above)
Comment (steuhs) [reply coverage/docs/design.md]: Done.

File: coverage/docs/design.md (hunk 32 above)
Comment (steuhs) [reply coverage/docs/design.md]: Done.

File: coverage/docs/design.md (hunk 31 above)
Comment (steuhs) [reply coverage/docs/design.md]: Done.

File: coverage/docs/design.md (hunk 30 above)
Comment (steuhs) [reply coverage/docs/design.md]: Done.

File: coverage/docs/design.md (hunk 29 above)
Comment (steuhs) [reply coverage/docs/design.md]: Done.

File: coverage/docs/design.md (hunk 28 above)
Comment (steuhs) [reply coverage/docs/design.md]: Done.

File: coverage/docs/design.md (hunk 27 above)
Comment (steuhs) [reply coverage/docs/design.md]: Done.

File: coverage/docs/design.md (hunk 26 above)
Comment (steuhs) [reply coverage/docs/design.md]: Done. Deleted

File: coverage/docs/design.md (hunk 25 above)
Comment (steuhs) [reply coverage/docs/design.md]: Done.

File: coverage/docs/design.md (hunk 24 above)
Comment (steuhs) [reply coverage/docs/design.md]: Done.

**Comment by steuhs** [comment]
@cjwagner @stevekuznetsov addressed all comments on design doc

**Comment by dims** [comment]
@steuhs do you have an example run(s) of the tool? wanted to peek at what we do cover today in our test harness.

**Comment by steuhs** [comment]
@dims Here is an example: https://github.com/knative/serving/pull/1833

**Comment by fejta** [comment]
Will you please link the design doc in the PR description? Also is there a reason this needs to be 36 commits?

**Comment by steuhs** [comment]
@fejta 
I am going to close this PR
Re-organized the PR into three commits here https://github.com/kubernetes/test-infra/pull/9313
Design doc in this PR https://github.com/kubernetes/test-infra/pull/9302
